src/libostree/ostree-repo-checkout.c

```diff
@@ -362,6 +375,12 @@ checkout_file_hardlink (OstreeRepo                          *self,
     {
       ret_was_supported = FALSE;
     }
+  else if (errno == EEXIST && options->overwrite_mode == OSTREE_REPO_CHECKOUT_OVERWRITE_ADD_FILES)
+    {
+      /* In this mode, we keep existing content.  Still need to mark the hardlink as supported.
+       */
+      ret_was_supported = TRUE;
```

Comment:
One concern with this is that the kernel checks for `EEXISTS` before checking whether hardlinking is supported. So we can't really assert this, right? Looking at `checkout_one_file_at()`, it relies on this info to update the devino cache, which will be wrong after this. Maybe add e.g. `out_exists` also so callers can distinguish this?